### PR TITLE
Explicit check for (USB) serial port disconnection (issue #22)

### PR DESCRIPTION
**Pull request by ypiguet-epfl**

### CODE REVIEW

File: dashel/dashel-win32.cpp [hunk 1]
```diff
@@ -933,13 +945,69 @@ namespace Dashel
...
+			if (hf == INVALID_HANDLE_VALUE)
+				throw DashelException(DashelException::ConnectionFailed, GetLastError(),
+					(std::string("Cannot open serial port ") + devName + ".").c_str());
 
 			buildDCB(hf, target.get<int>("baud"), target.get<int>("bits"), target.get<bool>("dtr"), target.get("parity"), target.get("stop"), target.get("fc"));
 
 			startStream(EvPotentialData);
 		}
+
+		virtual bool checkConnection()
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: If this function does not affect the object (i.e. it is only a check), it should be marked `const`.

File: dashel/dashel-win32.cpp [hunk 2]
```diff
@@ -805,6 +806,10 @@ namespace Dashel
 	//! Serial port stream
 	class SerialStream : public FileStream
 	{
+	protected:
+		std::string devName;
+		bool disconnected;
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: This member is unused, please remove.

File: dashel/dashel-win32.cpp [hunk 3]
```diff
@@ -933,13 +945,69 @@ namespace Dashel
...
+			deviceInfoData.cbSize = sizeof(deviceInfoData);
+			HDEVINFO hDevInfo = SetupDiGetClassDevs(&GUID_DEVCLASS_PORTS, 0, 0, DIGCF_PRESENT);
+			for (int i = 0; SetupDiEnumDeviceInfo(hDevInfo, i, &deviceInfoData); i++)
+			{
+				DWORD dataType;
+				DWORD bufSize;
+				TCHAR *buffer;
+
+				SetupDiGetDeviceRegistryProperty(hDevInfo, &deviceInfoData, SPDRP_FRIENDLYNAME, &dataType, NULL, 0, &bufSize);
+				buffer = new TCHAR[bufSize];
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: Please use a `std::vector<TCHAR>`, as this is exception-safe.

File: dashel/dashel-win32.cpp [hunk 4]
```diff
@@ -933,13 +945,69 @@ namespace Dashel
...
+								|| !(pulStatus & DN_WILL_BE_REMOVED);
+							delete[] buffer;
+							return connected;
+						}
+					}
+				}
+				delete[] buffer;
+			}
+
+			// device not found by SetupDiEnumDeviceInfo, assume it's disconnected
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: Grammar error "it's disconnected" => "it is disconnected".

File: dashel/dashel-win32.cpp [hunk 5]
```diff
@@ -1379,7 +1447,7 @@ namespace Dashel
 		
 		// Loop in order to consume all events, mostly within lock, excepted for wait
 		lock();
-		do
+		while (true)
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: Is there a reason to change the `do { ... } while(true)` into a `while (true) { ... };` excepted aesthetic? If so, I suggest to cancel this change to make the patch as readable as possible.

File: dashel/dashel-win32.cpp [hunk 6]
```diff
@@ -805,6 +806,9 @@ namespace Dashel
 	//! Serial port stream
 	class SerialStream : public FileStream
 	{
+	protected:
+		std::string devName;
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: Please document for Doxygen (for instance with //< ... after the member).

File: dashel/dashel-win32.cpp [hunk 7]
```diff
@@ -884,6 +888,14 @@ namespace Dashel
 			return true;
 		}
 
+		static std::string devNameToPortName(std::string const &devName)
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: Please briefly document the function for Doxygen.

File: dashel/dashel-win32.cpp [hunk 8]
```diff
@@ -933,13 +944,66 @@ namespace Dashel
...
+			if (hf == INVALID_HANDLE_VALUE)
+				throw DashelException(DashelException::ConnectionFailed, GetLastError(),
+					(std::string("Cannot open serial port ") + devName + ".").c_str());
 
 			buildDCB(hf, target.get<int>("baud"), target.get<int>("bits"), target.get<bool>("dtr"), target.get("parity"), target.get("stop"), target.get("fc"));
 
 			startStream(EvPotentialData);
 		}
+
+		virtual bool checkConnection() const
```
Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: Why is it virtual? It only applies to `SerialStream`, right?

Comment (ypiguet-epfl) [inline dashel/dashel-win32.cpp]: Likely to be overridden should the SerialStream class (not declared as final) be subclassed

Comment (stephanemagnenat) [inline dashel/dashel-win32.cpp]: I see. In general, the design decision behind my software is to have the least possible virtuals, at least in non-public API. As this is a private API (SerialStream is never exposed to the user of the library), I suggest to keep it non-virtual, as we could just add the virtual in the hypothetical case we want to override it.

**Comment by stephanemagnenat** [comment]
The code is ready to merge on my side. Please rebase in a single commit, whose message refer the problem and the issue number it fixes (#11). Thank you very much!

**Pull request merged**